src/Catalog/PackageCatalogItemCreator.cs

```diff
@@ -0,0 +1,234 @@
+            }
+
+            return new PackageCatalogItemCreator(httpClient, telemetryService, logger, storage);
+        }
+
+        public async Task<PackageCatalogItem> CreateAsync(
+            FeedPackageDetails packageItem,
+            DateTime timestamp,
+            CancellationToken cancellationToken)
+        {
+            if (packageItem == null)
+            {
+                throw new ArgumentNullException(nameof(packageItem));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            PackageCatalogItem item = null;
+
+            if (_storage != null)
+            {
+                item = await GetPackageViaStorageAsync(packageItem, cancellationToken);
+            }
+
+            if (item == null)
+            {
+                item = await GetPackageViaHttpAsync(packageItem, timestamp, item, cancellationToken);
+            }
+
+            return item;
+        }
+
+        private async Task<PackageCatalogItem> GetPackageViaStorageAsync(FeedPackageDetails packageItem, CancellationToken cancellationToken)
+        {
+            PackageCatalogItem item = null;
+            var packageName = $"{packageItem.PackageId.ToLowerInvariant()}.{packageItem.PackageVersion.ToLowerInvariant()}.nupkg";
+            var blob = _storage.GetCloudBlockBlobReference(packageName);
+
+            if (blob == null)
+            {
+                _telemetryService.TrackMetric(
+                    TelemetryConstants.NonExistentBlob,
+                    metric: 1,
+                    properties: GetProperties(packageItem, blob));
+            }
+            else
+            {
+                await blob.FetchAttributesAsync(cancellationToken);
+
+                string packageHash = null;
+                var etag = blob.ETag;
+
+                var metadata = await blob.GetMetadataAsync(cancellationToken);
+
+                if (metadata.TryGetValue("SHA512", out packageHash))
+                {
+                    using (var stream = await blob.GetStreamAsync(cancellationToken))
+                    {
+                        item = Utils.CreateCatalogItem(
+                            packageItem.ContentUri.ToString(),
+                            stream,
+                            packageItem.CreatedDate,
+                            packageItem.LastEditedDate,
+                            packageItem.PublishedDate,
+                            licenseNames: null,
+                            licenseReportUrl: null,
+                            packageHash: packageHash);
+
+                        if (item == null)
+                        {
+                            _logger.LogWarning("Unable to extract metadata from: {PackageDetailsContentUri}", packageItem.ContentUri);
+                        }
+                    }
+
+                    if (item != null)
+                    {
+                        await blob.FetchAttributesAsync(cancellationToken);
```

Comment:
As a side note (not explicitly related to this PR), what happens if:
1. A package is pushed.
2. Every V3 job except Feed2Catalog stops running.
3. Feed2Catalog commits this package to the catalog.
4. The package is deleted.
5. Feed2Catalog commits this package's delete to the catalog.
6. The V3 jobs start running again and process the package's creation and deletion commits.

Have we verified that all jobs that download the package from the packages container skip a commit if the package associated with that commit doesn't exist in the packages container?

Reply:
That question is unrelated to this PR.  :)